library/ckagen.cpp

```diff
@@ -0,0 +1,363 @@
+#include "ckagen.h"
+#include <algorithm>
+#include <mpi.h>
+#include "kagen.h"
+
+struct kagen_gen {
+    void* gen_ptr;
```

Comment:
Since this is only used in the cpp file, why not use the right pointer type instead of `void *`?